cordova-plugin-outline/android/java/org/outline/OutlinePlugin.java

```diff
@@ -288,135 +294,86 @@ public void onActivityResult(int request, int result, Intent data) {
       LOG.warning("Received non-requested activity result.");
       return;
     }
-    if (result != RESULT_OK) {
-      LOG.severe("Failed to prepare VPN.");
-      PluginResult pluginResult =
-          new PluginResult(PluginResult.Status.ERROR, ErrorCode.VPN_PERMISSION_NOT_GRANTED.value);
-      sendPluginResult(startRequestTunnelId, Action.START.value, pluginResult, false);
-      return;
-    }
-    try {
-      startVpnTunnel();
-    } catch (Exception e) {
-      // Do not propagate the exception in the main thread; the service will broadcast the error.
-    }
-  }
-
-  private void startVpnTunnel() {
-    LOG.info("Starting VPN Tunnel");
-    if (vpnTunnelService == null) {
-      onVpnTunnelServiceNotBound(Action.START, startRequestTunnelId);
+    if (result != Activity.RESULT_OK) {
+      LOG.warning("Failed to prepare VPN.");
+      sendErrorCode(startVpnRequest.callback, ErrorCode.VPN_PERMISSION_NOT_GRANTED.value);
       return;
     }
-    vpnTunnelService.startTunnel(startRequestTunnelId, startRequestConfig);
+    executeAsync(Action.START.value, startVpnRequest.args, startVpnRequest.callback);
+    startVpnRequest = null;
   }
 
-  private void stopVpnTunnel(final String tunnelId) {
-    LOG.info("Stopping VPN tunnel.");
-    if (vpnTunnelService == null) {
-      onVpnTunnelServiceNotBound(Action.STOP, tunnelId);
-      return;
+  private int startVpnTunnel(final String tunnelId, final JSONObject config) throws Exception {
+    LOG.info(String.format(Locale.ROOT, "Starting VPN tunnel %s", tunnelId));
+    final TunnelConfig tunnelConfig;
+    try {
+      tunnelConfig = VpnTunnelService.makeTunnelConfig(tunnelId, config);
+    } catch (Exception e) {
+      LOG.log(Level.SEVERE, "Failed to retrieve the tunnel proxy config.", e);
+      return ErrorCode.ILLEGAL_SERVER_CONFIGURATION.value;
     }
-    vpnTunnelService.stopTunnel(tunnelId);
+    return vpnTunnelService.startTunnel(tunnelConfig);
   }
 
   // Returns whether the VPN service is running a particular tunnel instance.
   private boolean isTunnelActive(final String tunnelId) {
-    if (vpnTunnelService == null) {
-      onVpnTunnelServiceNotBound(Action.IS_RUNNING, tunnelId);
-      return false;
+    boolean isActive = false;
+    try {
+      isActive = vpnTunnelService.isTunnelActive(tunnelId);
```

Comment:
No need for `isActive`.
```suggestion
      return vpnTunnelService.isTunnelActive(tunnelId);
```

Reply:
Done in 86eea94